infra/gravity/provision.go

```diff
@@ -0,0 +1,257 @@
+package gravity
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/robotest/infra"
+	"github.com/gravitational/robotest/infra/terraform"
+	sshutil "github.com/gravitational/robotest/lib/ssh"
+
+	"github.com/gravitational/trace"
+
+	"github.com/go-yaml/yaml"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/go-playground/validator.v9"
+)
+
+type ProvisionerConfig struct {
+	// DeployTo defines cloud to deploy to
+	CloudProvider string `yaml:"cloud_provider" validate:"required,eq=aws|eq=azure"`
+	// AWS defines AWS connection parameters
+	AWS *infra.AWSConfig `yaml:"aws"`
+	// Azure defines Azure connection parameters
+	Azure *infra.AzureConfig `yaml:"azure"`
+
+	// ScriptPath is the path to the terraform script or directory for provisioning
+	ScriptPath string `yaml:"script_path" validate:"required"`
+	// InstallerURL is AWS S3 URL with the installer
+	InstallerURL string `yaml:"installer_url" validate:"required,url`
+}
+
+func LoadConfig(t *testing.T, configFile string) *ProvisionerConfig {
+	require.NotEmpty(t, configFile, "config file")
+	f, err := os.Open(configFile)
+	require.NoError(t, err, configFile)
+	defer f.Close()
+
+	configBytes, err := ioutil.ReadAll(f)
+	require.NoError(t, err)
+
+	cfg := ProvisionerConfig{}
+	err = yaml.Unmarshal(configBytes, &cfg)
+	require.NoError(t, err, configFile)
+
+	err = validator.New().Struct(&cfg)
+	if err == nil {
+		return &cfg
+	}
+
+	if validationErrors, ok := err.(validator.ValidationErrors); ok {
+		t.Errorf("Config %s has errors:", configFile)
+		for _, fieldError := range validationErrors {
+			t.Errorf(" * %s=\"%v\" fails \"%s\"", fieldError.Field(), fieldError.Value(), fieldError.Tag())
+		}
+		require.FailNow(t, "Fix config")
+	}
+
+	// never reached
+	return nil
+}
+
+// function which will destroy previously created remote resources
+type DestroyFn func() error
+
+// scheduleDestroy will register resource (placement) group for destruction with external service
+// based on Context expiration deadline. This would only be used in continuous operation (i.e. part of CLI)
+func scheduleDestroy(ctx context.Context, cloud, tag string) {
+	// TODO: implement
+}
+
+// cloudDynamicParams is a necessary evil to marry terraform vars, e2e legacy objects and needs of this provisioner
+type cloudDynamicParams struct {
+	dockerDevice string
+	user         string
+	os           string
+	installDir   string
+	installerUrl string
+	tf           terraform.Config
+	env          map[string]string
+}
+
+// makeDynamicParams takes base config, validates it and returns TerraformConfig
```

Comment:
I meant that that the original comment states it returns a TerraformConfig while it does not:
`// makeDynamicParams takes base config, validates it and returns TerraformConfig`